contracts/TaskTreasury/TaskTreasuryUpgradable.sol

```diff
@@ -0,0 +1,230 @@
+// SPDX-License-Identifier: UNLICENSED
+pragma solidity 0.8.11;
+
+import {
+    EnumerableSet
+} from "@openzeppelin/contracts/utils/structs/EnumerableSet.sol";
+import {
+    SafeERC20,
+    IERC20
+} from "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+import {
+    OwnableUpgradeable
+} from "@openzeppelin/contracts-upgradeable/access/OwnableUpgradeable.sol";
+import {
+    ReentrancyGuardUpgradeable
+} from "@openzeppelin/contracts-upgradeable/security/ReentrancyGuardUpgradeable.sol";
+import {
+    Initializable
+} from "@openzeppelin/contracts-upgradeable/proxy/utils/Initializable.sol";
+import {Math} from "@openzeppelin/contracts/utils/math/Math.sol";
+import {_transfer, ETH} from "../vendor/gelato/FGelato.sol";
+import {ITaskTreasury} from "../interfaces/ITaskTreasury.sol";
+
+contract TaskTreasuryUpgradable is
+    Initializable,
+    OwnableUpgradeable,
+    ReentrancyGuardUpgradeable
+{
+    using EnumerableSet for EnumerableSet.AddressSet;
+    using SafeERC20 for IERC20;
+
+    address payable public immutable gelato;
+    ITaskTreasury public immutable oldTreasury;
+
+    mapping(address => mapping(address => uint256)) public userTokenBalance;
+    mapping(address => EnumerableSet.AddressSet) internal _tokenCredits;
+    EnumerableSet.AddressSet internal _whitelistedServices;
+
+    event FundsDeposited(
+        address indexed sender,
+        address indexed token,
+        uint256 indexed amount
+    );
+    event FundsWithdrawn(
+        address indexed receiver,
+        address indexed initiator,
+        address indexed token,
+        uint256 amount
+    );
+
+    event LogDeductFees(
+        address indexed user,
+        address indexed executor,
+        address indexed token,
+        uint256 fees,
+        address service
+    );
+
+    modifier onlyWhitelistedServices() {
+        require(
+            _whitelistedServices.contains(msg.sender),
+            "TaskTreasuryAccounting: onlyWhitelistedServices"
+        );
+        _;
+    }
+
+    constructor(address payable _gelato, address _oldTreasury) {
```

Comment:
@brandonchuah in a proxy contract, you need to execute constructor logic in the `initialize` method, otherwise the state is manipulated on the implementation contract not the proxy contract.

Reply:
would be beneficial to include test to check inital state of contract after deployment.

Reply:
Those variables are immutables and can be initialized in the constructor right?

Reply:
Is it possible to write a test case to check the state on the proxy after deployment?

Reply:
added in c1056a1